core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/score/director/violation/SolutionSnapshot.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.director.violation;
+
+import static java.util.Comparator.comparing;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.optaplanner.core.impl.domain.solution.descriptor.SolutionDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.ShadowVariableDescriptor;
+
+/**
+ * Serves for detecting shadow variables' corruption. When the snapshot is created, it records the state of all shadow variables
+ * of all entities. The {@link #createShadowVariablesViolationMessage} method takes a look at the shadow variables again,
+ * compares their state with the recorded one and describes the difference in a violation message.
+ */
+public final class SolutionSnapshot {
```

Comment:
The code is good.
Just wondering about naming:
This class SolutionSnapshot only works for shadow variables (corruption) - shouldn't the name convey that?

I get the impression that this entire code could be the responsibility of VariableListenerSupport? (Clearly, AbstractScoreDirector was too bloated to get this responsibility). Wdyt?

Reply:
> This class SolutionSnapshot only works for shadow variables (corruption) - shouldn't the name convey that?

I think the class name is good, I would rather keep it simple until there are more types snapshots. As long as there is only one, no need to fully describe what the class does in its name. I'm open to discussion if you offer some alternatives.

However, I based on this suggestion, I could see this as an improvement:
```diff
- SolutionSnapshot solutionSnapshot = SolutionSnapshot.of(getSolutionDescriptor(), workingSolution);
+ SolutionSnapshot solutionSnapshot = SolutionSnapshot.captureShadowVariables(getSolutionDescriptor(), workingSolution);
```
Wdyt?